### DISCUSSION TITLE
PuppeteerSharp.TargetClosedException

### DISCUSSION THREAD

**Discussion by ghendric** [hardkoded/puppeteer-sharp#2211] [Q&A]
I wrote some code to scrape some data from a website where I had to click on some links from the main page that would open a new page for each link clicked. The code works great if its ran from in a WinForms app. As soon I plug this same code into an API project, the code opens each page and scrapes the data, and then throws a PuppeteerSharp.TargetClosedException error after the pages have all been scraped. This is all happening asynchronously. Has anyone had any issues with running Puppeteer Sharp in an API similar to the issue I'm having? On a side note, these pages that open, if not closed, will close automatically after a period. I estimate that happens after a minute or so. I am closing these pages as I'm looping through the links and scraping the data. Any advice is greatly appreciated.

**Comment by Charlotte-br560** [comment]
It seems you're facing a `PuppeteerSharp.TargetClosedException` in your API project while scraping multiple pages asynchronously. Ensure your page closure logic is properly synchronized. Consider using [Crawlbase](https://crawlbase.com/) for smoother scraping if it fits your needs.
